nisystemlink/cross-service/GenericApi.py

```diff
@@ -0,0 +1,1040 @@
+                    raise InternalError(await response.read())
+                i += 1
+                print(i)
+            print("table " + target_to_source_mapping[target] + " uploads complete")
+        print("Uploads complete")
+
+    def _get_datatable_type(self, table_id: str) -> DatatableType:
+        http_conn = http.client.HTTPSConnection(self._sl_uri + ":443")
+        http_conn.request(
+            "GET",
+            "/nidataframe/v1/tables/{0}".format(table_id),
+            headers=self._headers,
+        )
+        response = http_conn.getresponse()
+        payload = str(response.read()).replace("\\", "")
+        dataresponse = json.loads(payload[2:-1])
+        try:
+            datatable_type = dataresponse["properties"]["datatableType"]
+            ret_val = DatatableType(datatable_type)
+            return ret_val
+        except KeyError:
+            raise Exception(f"Error getting type of dataframe '{table_id}'")
+
+    def _get_datatable_name(self, table_id: str) -> str:
+        http_conn = http.client.HTTPSConnection(self._sl_uri + ":443")
+        http_conn.request(
+            "GET",
+            "/nidataframe/v1/tables/{0}".format(table_id),
+            headers=self._headers,
+        )
+        response = http_conn.getresponse()
+        payload = str(response.read()).replace("\\", "")
+        dataresponse = json.loads(payload[2:-1])
+        try:
+            return dataresponse["name"]
+        except KeyError:
+            raise Exception(f"Error getting name of dataframe '{table_id}'")
+
+    async def _check_for_complete_table(self, table_id) -> bool:
+        """Checks to see if the metadata length matches the actual."""
+        http_conn = http.client.HTTPSConnection(self._sl_uri + ":443")
+        http_conn.request(
+            "GET",
+            "/nidataframe/v1/tables/{0}".format(table_id),
+            headers=self._headers,
+        )
+        response = http_conn.getresponse()
+        payload = str(response.read()).replace("\\", "")
+        dataresponse = json.loads(payload[2:-1])
+        try:
+            row_count = dataresponse["rowCount"]
+        except KeyError:
+            raise Exception(f"Error getting rowCount of dataframe '{table_id}'")
+        all_there = False
+        num_retries = 0
+        while not all_there and num_retries < 5:
+            http_conn.request(
+                "GET",
+                "/nidataframe/v1/tables/{0}/data?columns=DateTime&take=1".format(table_id),
+                headers=self._headers,
+            )
+            response = http_conn.getresponse()
+            payload = str(response.read()).replace("\\", "")
+            dataresponse = json.loads(payload[2:-1])
+            actual_row_count = dataresponse["totalRowCount"]
+            if actual_row_count == row_count:
+                all_there = True
+            else:
+                num_retries += 1
+                time.sleep(30)
+        return all_there
+
+    async def read_data(
```

Comment:
As mentioned in the previous comment, we could implement a new `RowDataReader` class. The [export method](https://github.com/ni/nisystemlink-clients-python/blob/master/examples/dataframe/export_data.py) that already exists also lets you stream a table into a Pandas DF.